src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/ExtendedQueryTagDataRowsBuilder.cs

```diff
@@ -182,17 +202,22 @@ private static void AddLongRow(DicomDataset instance, List<InsertLongExtendedQue
 
             if (longVal.HasValue)
             {
-                longRows.Add(new InsertLongExtendedQueryTagTableTypeV1Row(queryTag.ExtendedQueryTagStoreEntry.Key, longVal.Value, (byte)queryTag.Level));
+                longRows.Add(new InsertLongExtendedQueryTagTableTypeV1Row(tagKey, longVal.Value, (byte)queryTag.Level));
             }
         }
 
-        private static void AddStringRow(DicomDataset instance, List<InsertStringExtendedQueryTagTableTypeV1Row> stringRows, QueryTag queryTag)
+        private static void AddStringRow(DicomDataset instance, List<InsertStringExtendedQueryTagTableTypeV1Row> stringRows, QueryTag queryTag, int tagKey)
         {
             string stringVal = instance.GetSingleValueOrDefault<string>(queryTag.Tag, expectedVR: queryTag.VR);
             if (stringVal != null)
             {
-                stringRows.Add(new InsertStringExtendedQueryTagTableTypeV1Row(queryTag.ExtendedQueryTagStoreEntry.Key, stringVal, (byte)queryTag.Level));
+                stringRows.Add(new InsertStringExtendedQueryTagTableTypeV1Row(tagKey, stringVal, (byte)queryTag.Level));
             }
         }
+
+        private static int GetKeyFromQueryTag(QueryTag queryTag)
+        {
+            return queryTag.WorkitemQueryTagStoreEntry != null ? queryTag.WorkitemQueryTagStoreEntry.Key : queryTag.ExtendedQueryTagStoreEntry.Key;
```

Comment:
Should this just be a property on the `QueryTag` class?

Reply:
Yes that is correct. we thought about that. It needs some refactoring and change existing code.